app/utility/base_knowledge_svc.py

```diff
@@ -0,0 +1,318 @@
+        """
+        Check whether or not a rule matches the provided criteria
+        :param rule: Rule object to check
+        :param desired_quals: dictionary containing fields to match on
+        :param wildcard: whether or not to do wildcard matching on the 'match' field
+        :return:
+        """
+        criteria_matches = []
+        for k, v in desired_quals.items():
+            if getattr(rule, k) == v:
+                criteria_matches.append(True)
+            elif k == 'match' and wildcard:  # support wildcard matching
+                val = getattr(rule, k)
+                local_stub = val.split('*')
+                input_stub = v.split('*')
+                if all(x in v for x in local_stub) or all(x in val for x in input_stub):
+                    criteria_matches.append(True)
+        if len(criteria_matches) == len(desired_quals):
+            return True
+
+    # --- New Inferencing API ---
+    # NOT IMPLEMENTED YET
+    def _similar_facts(self, fact, agent, group):
+        """return facts that are close to supplied fact.
+
+
+        Ex:
+            - other facts for an agent with given trait/value
+            - other facts for the group/agent
+            - other facts with same value
+        """
+        raise NotImplementedError
+
+    def _fact_value_distribution(self, critera, agent=None, group=None):
+        """return the value distribution for the given fact, and further filtered down
+        to agent/group if supplied
+
+
+        Ex: fact value distribution for 'host.user.name' on group 'workstations':
+            --> [{'admin': .4}, {'system': .4}, {'michael': .1}, {'workstation1': .1}]
+        """
+        raise NotImplementedError
+
+    def _best_guess(self, criteria, agent=None, group=None):
+        """wrapper around 'fact_value_distribution', just returning highest probable value"""
+        raise NotImplementedError
+
+    def _best_facts(self, agent=None, group=None, metric='usage_success'):
+        """best facts based on requested metric
+
+
+        Args:
+            metric: ['usage_success', 'most_recent_success', ...]
+        """
+        raise NotImplementedError
+
+    def _locate(self, object_name, query=None):
+        """
+        Locate a matching object in the internal store
+        :param object_name: object type
+        :param query: dictionary of fields to match on
+        :return: list of matching objects
+        """
+        try:
+            return [obj for obj in self.fact_ram[object_name] if obj.match(query)]
+        except Exception as e:
+            self.log.error('[!] LOCATE: %s' % e)
+
+    def _remove(self, object_name, query):
+        """
+        Remove objects from the internal store
+        :param object_name: object type
+        :param query: dictionary of fields to match on
+        """
+        try:
+            sublist = self.fact_ram[object_name] = [obj for obj in self.fact_ram[object_name] if obj.match(query)]
```

Comment:
This looks erroneous. 

Would cause L235 to always to set objects in ram to empty list; i.e remove everything. Right?

Reply:
I have no idea what the heck happened there - it should just be `sublist = [obj..`. The `= self.fact_ram... =` is definitely not supposed to be there. I've fixed it and updated one of the removal tests to check for proper, singular, deletion.